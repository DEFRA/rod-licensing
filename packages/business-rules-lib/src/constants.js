/**
 * The maximum days in advance of the purchase date that the licence may start
 * @type {number}
 */
export const ADVANCED_PURCHASE_MAX_DAYS = 30

/**
 * The  maximum number of permissions in a (multi-buy) transaction
 * @type {number}
 */
export const MAX_PERMISSIONS_PER_TRANSACTION = 50

/**
 * Allowed status values for the payment journal
 */
export const PAYMENT_JOURNAL_STATUS_CODES = {
  InProgress: 'In Progress',
  Cancelled: 'Cancelled',
  Failed: 'Failed',
  Expired: 'Expired',
  Completed: 'Completed'
}

/**
 * Payment status returned by the GOV.UK pay API
 */
export const PAYMENT_STATUS = {
  Success: 'success',
  Failure: 'failed'
}

/**
 * Failure states returned by the GOV.UK pay API
 * https://docs.payments.service.gov.uk/api_reference/#errors-caused-by-payment-statuses
 */
export const GOVUK_PAY_ERROR_STATUS_CODES = {
  REJECTED: 'P0010',
  EXPIRED: 'P0020',
  USER_CANCELLED: 'P0030',
  ERROR: 'P0050',
  NOT_FOUND: 'P0200'
}

/**
 * Constants used in finalization
 * @type {{telesales: string, govPay: string}}
 */
export const TRANSACTION_SOURCE = { govPay: 'Gov Pay', telesales: 'Telesales' }
export const PAYMENT_TYPE = { debit: 'Debit card' }

/**
 * Data sources associated with POCL transactions
 */
export const POCL_DATA_SOURCE = 'Post Office Sales'
export const DDE_DATA_SOURCE = 'DDE File'
export const POCL_TRANSACTION_SOURCES = [POCL_DATA_SOURCE, DDE_DATA_SOURCE]

/**
 * Describes the lifetime of the renewal link.
 */
export const RENEW_BEFORE_DAYS = 60 // The number of days before licence expiry the link is active
export const RENEW_AFTER_DAYS = 60 // The number of days after licence expiry the link is active

/**
 * The number of minutes after payment that a licence will start
 * @type {number}
 */
export const START_AFTER_PAYMENT_MINUTES = 30

/**
 * Timezone of the service
 */
export const SERVICE_LOCAL_TIME = 'Europe/London'

/**
 * Date for switching paper fulfilment provider
 */
export const FULFILMENT_SWITCHOVER_DATE = '2024-05-30T23:00:00.000Z'
