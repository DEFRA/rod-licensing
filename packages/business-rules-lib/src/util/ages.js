/** The maximum age at which an angler is considered to be a minor (free licence) */
export const MINOR_MAX_AGE = 12
/** The maximum age at which an angler is entitled to a junior concession */
export const JUNIOR_MAX_AGE = 16
/** The minimum age at which an angler becomes entitled to a senior concession */
export const SENIOR_MIN_AGE = 66

/**
 * Determine if the provided age is classified as a minor
 * @param {number} age The age to be tested
 * @returns {boolean} true if the given age should be classified as a minor
 */
export const isMinor = age => age <= MINOR_MAX_AGE

/**
 * Determine if the provided age is classified as a junior
 * @param {number} age The age to be tested
 * @returns {boolean} true if the given age should be classified as a junior
 */
export const isJunior = age => age > MINOR_MAX_AGE && age <= JUNIOR_MAX_AGE

/**
 * Determine if the provided age is classified as a senior
 * @param {number} age The age to be tested
 * @returns {boolean} true if the given age should be classified as a senior
 */
export const isSenior = age => age >= SENIOR_MIN_AGE
