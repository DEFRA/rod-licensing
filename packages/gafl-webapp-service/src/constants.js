/**
 * System constants and defaults
 */
export const ADDRESS_LOOKUP_SERVICE = { lang: 'EN', dataset: 'DPA' }
export const ADDRESS_LOOKUP_TIMEOUT_MS_DEFAULT = 10000
export const ATTRIBUTION_REDIRECT_DEFAULT = '/'
export const SESSION_TTL_MS_DEFAULT = 3 * 60 * 60 * 1000
export const PORT_DEFAULT = 3000
export const REDIS_PORT_DEFAULT = 6379
export const PAGE_STATE = { completed: true, error: false }
export const SESSION_COOKIE_NAME_DEFAULT = 'sid'
export const CSRF_TOKEN_COOKIE_NAME_DEFAULT = 'rlsctkn'
export const FEEDBACK_URI_DEFAULT = '#'
export const CHANNEL_DEFAULT = 'websales'
export const SERVICE_PAGE_DEFAULT = 'https://www.gov.uk/fishing-licences/buy-a-fishing-licence'

export const COMPLETION_STATUS = {
  agreed: 'agreed',
  posted: 'posted',
  paymentCreated: 'payment-created',
  paymentCancelled: 'payment-cancelled',
  paymentFailed: 'payment-failed',
  paymentCompleted: 'payment-completed',
  finalised: 'finalised',
  completed: 'completed'
}

export const GOVPAYFAIL = {
  prePaymentRetry: { step: 'pre-payment' },
  postPaymentRetry: { step: 'post-payment' }
}

export const UTM = {
  CAMPAIGN: 'utm_campaign',
  MEDIUM: 'utm_medium',
  SOURCE: 'utm_source',
  TERM: 'utm_term',
  CONTENT: 'utm_content'
}

export const dateFormats = ['YYYY-MM-DD', 'YY-MM-DD', 'YYYY-M-DD', 'YY-M-DD', 'YYYY-MM-D', 'YY-MM-D', 'YYYY-M-D', 'YY-M-D']

export const CommonResults = {
  SUMMARY: 'summary',
  OK: 'ok',
  NO: 'no',
  YES: 'yes'
}

export const allowsPhysicalLicence = {
  YES: 'yes',
  NO: 'no'
}

// If the user has seen the summary page these are set in the status
export const CONTACT_SUMMARY_SEEN = 'contact-summary'
export const LICENCE_SUMMARY_SEEN = 'licence-summary'

// These cookies are used by the load balancer
export const ALB_COOKIE_NAME = 'AWSALBTG'
export const ALBCORS_COOKIE_NAME = 'AWSALBTGCORS'
