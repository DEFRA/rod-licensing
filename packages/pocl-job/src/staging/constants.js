export const MAX_CREATE_TRANSACTION_BATCH_SIZE = 25
export const MAX_FINALISE_TRANSACTION_BATCH_SIZE = 10

export const POST_OFFICE_DATASOURCE = 'Post Office Sales'
export const DIRECT_DEBIT_DATASOURCE = 'DDE File'
export const DIRECT_DEBIT_PAYMENTSOURCE = 'Direct Debit'
export const POSTAL_ORDER_DATASOURCE = 'Postal Order Sales'
export const POSTAL_ORDER_PAYMENTSOURCE = 'Postal Order'

export const FILE_STAGE = {
  Pending: 'Received and Pending',
  Staging: 'Staging Transactions',
  Finalising: 'Finalising Transactions',
  Completed: 'Completed'
}

export const DYNAMICS_IMPORT_STAGE = {
  Pending: 'Received and Pending',
  InProgress: 'In Progress',
  Processed: 'Processed',
  ProcessedWithWarnings: 'Processed with Warnings',
  Failed: 'Failed',
  UnderReview: 'Under Review',
  Closed: 'Closed',
  isAlreadyProcessed: status => status !== DYNAMICS_IMPORT_STAGE.Pending && status !== DYNAMICS_IMPORT_STAGE.InProgress
}

export const RECORD_STAGE = {
  TransactionCreated: 'Transaction Created',
  TransactionCreationFailed: 'Transaction Creation Failed',
  TransactionFinalised: 'Transaction Finalised',
  TransactionFinalisationFailed: 'Transaction Finalisation Failed'
}
