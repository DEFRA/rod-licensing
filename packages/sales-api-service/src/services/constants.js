const CONCESSION = {
  SENIOR: 'Senior',
  JUNIOR: 'Junior',
  DISABLED: 'Disabled'
}

const CONCESSION_PROOF = {
  none: 'No Proof',
  blueBadge: 'Blue Badge',
  NI: 'National Insurance Number'
}

export { CONCESSION, CONCESSION_PROOF }
