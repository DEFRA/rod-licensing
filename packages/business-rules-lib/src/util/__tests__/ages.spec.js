import { CONCESSION, CONCESSION_PROOF, HOW_CONTACTED } from '../../constants.js'
import moment from 'moment-timezone'
import * as ages from '../ages.js'

const permission = {}

const junior = {
  type: CONCESSION.JUNIOR,
  proof: {
    type: CONCESSION_PROOF.none
  }
}

const senior = {
  type: CONCESSION.SENIOR,
  proof: {
    type: CONCESSION_PROOF.none
  }
}

const disabledbb = {
  type: CONCESSION.DISABLED,
  proof: {
    type: CONCESSION_PROOF.blueBadge,
    referenceNumber: '123'
  }
}

const disabledNi = {
  type: CONCESSION.DISABLED,
  proof: {
    type: CONCESSION_PROOF.NI,
    referenceNumber: '456'
  }
}

const LICENCE_START_DATE = '2020-06-06'

const getLicensee = ({ age, referenceDate = LICENCE_START_DATE } = {}) => ({
  firstName: 'Johnny',
  lastName: 'Test',
  preferredMethodOfConfirmation: HOW_CONTACTED.letter,
  birthDate: moment(referenceDate).subtract(age, 'years').format('YYYY-MM-DD')
})

describe('The concession helper', () => {
  describe('age determination', () => {
    describe('isMinor', () => {
      it.each(Array.from({ length: ages.MINOR_MAX_AGE }, (_v, index) => index + 1))('age of %d is a minor', age => {
        expect(ages.isMinor(age)).toBeTruthy()
      })

      it(`${ages.MINOR_MAX_AGE + 1} is not a minor`, () => {
        expect(ages.isMinor(ages.MINOR_MAX_AGE + 1)).toBeFalsy()
      })
    })

    describe('isJunior', () => {
      it.each(Array.from({ length: ages.JUNIOR_MAX_AGE - ages.MINOR_MAX_AGE }, (_v, index) => ages.MINOR_MAX_AGE + index + 1))(
        'Age of %d is a Junior',
        age => {
          expect(ages.isJunior(age)).toBeTruthy()
        }
      )
      it.each([ages.MINOR_MAX_AGE, ages.JUNIOR_MAX_AGE + 1])('Age of %d is not a junior', age => {
        expect(ages.isJunior(age)).toBeFalsy()
      })
    })

    describe('isSenior', () => {
      it.each(Array.from({ length: 10 }, (_v, index) => ages.SENIOR_MIN_AGE + index))('age of %d is a senior', age => {
        expect(ages.isSenior(age)).toBeTruthy()
      })

      it(`age of ${ages.SENIOR_MIN_AGE - 1} is not a senior`, () => {
        expect(ages.isSenior(ages.SENIOR_MIN_AGE - 1)).toBeFalsy()
      })
    })
  })

  describe('addJunior', () => {
    it('concessions contains junior', () => {
      ages.addJunior(permission)
      expect(permission.concessions).toContainEqual(junior)
    })

    it('concessions contains only 1 junior', () => {
      ages.addJunior(permission)
      expect(permission.concessions.length).toBe(1)
    })
  })

  describe('addSenior', () => {
    it('concessions contains senior', () => {
      ages.addSenior(permission)
      expect(permission.concessions).toContainEqual(senior)
    })

    it('concessions contains only 1 senior', () => {
      ages.addSenior(permission)
      expect(permission.concessions.length).toBe(1)
    })
  })

  describe('addDisabled blue badge', () => {
    it('concessions contains disabled blue badge', () => {
      ages.addDisabled(permission, CONCESSION_PROOF.blueBadge, '123')
      expect(permission.concessions).toContainEqual(disabledbb)
    })

    it('concessions contains only 2 disabled blue badge', () => {
      ages.addDisabled(permission, CONCESSION_PROOF.blueBadge, '123')
      expect(permission.concessions.length).toBe(2)
    })
  })

  describe('addDisabled NI', () => {
    it('concessions contains disabled NI', () => {
      ages.addDisabled(permission, CONCESSION_PROOF.NI, '456')
      expect(permission.concessions).toContainEqual(disabledNi)
    })

    it('concessions contains only 2 disabled NI', () => {
      ages.addDisabled(permission, CONCESSION_PROOF.NI, '456')
      expect(permission.concessions.length).toBe(2)
    })
  })

  it('add junior to disabled concessions contains junior', () => {
    ages.addJunior(permission)
    expect(permission.concessions).toContainEqual(junior)
  })

  it('add junior to disabled concessions contains only 2', () => {
    ages.addJunior(permission)
    expect(permission.concessions.length).toBe(2)
  })

  it('add NI to junior concessions contains junior', () => {
    ages.addDisabled(permission, CONCESSION_PROOF.NI, '456')
    expect(permission.concessions).toContainEqual(disabledNi)
  })

  it('add NI to junior concessions contains only 2 disabled NI', () => {
    ages.addDisabled(permission, CONCESSION_PROOF.NI, '456')
    expect(permission.concessions.length).toBe(2)
  })

  it('add senior to NI contains senior', () => {
    ages.addSenior(permission)
    expect(permission.concessions.length).toBe(2)
  })

  it('add senior to NI has 2 concessions', () => {
    ages.addSenior(permission)
    expect(permission.concessions.length).toBe(2)
  })

  it('add junior to replace senior contains junior', () => {
    ages.removeDisabled(permission)
    ages.addJunior(permission)
    expect(permission.concessions.length).toBe(1)
  })

  it('add junior to replace senior contains only 1 concession', () => {
    ages.removeDisabled(permission)
    ages.addJunior(permission)
    expect(permission.concessions.length).toBe(1)
  })

  it('add NI to replace senior contains ni', () => {
    ages.clear(permission)
    ages.addDisabled(permission, CONCESSION_PROOF.NI, '456')
    expect(permission.concessions).toContainEqual(disabledNi)
  })

  it('add NI to replace senior contains only 1 concession', () => {
    ages.clear(permission)
    ages.addDisabled(permission, CONCESSION_PROOF.NI, '456')
    expect(permission.concessions.length).toBe(1)
  })

  it('add senior to replace NI contains senior', () => {
    ages.clear(permission)
    ages.addSenior(permission)
    expect(permission.concessions).toContainEqual(senior)
  })

  it('add senior to replace NI has only 1 concession', () => {
    ages.clear(permission)
    ages.addSenior(permission)
    expect(permission.concessions.length).toBe(1)
  })

  describe('getAgeConcession', () => {
    ;[junior, senior].forEach(concession => {
      it(`returns the ${concession.type} concession if present`, () => {
        const permission = { concessions: [concession] }
        expect(ages.getAgeConcession(permission)).toBe(concession)
      })
    })

    it('returns undefined if only disabled concession present', () => {
      const permission = { concessions: [disabledNi] }
      expect(ages.getAgeConcession(permission)).toBe(undefined)
    })

    it("returns undefined if concessions don't exist", () => {
      const permission = {}
      expect(ages.getAgeConcession(permission)).toBe(undefined)
    })
  })

  describe('ageConcessionHelper', () => {
    beforeEach(jest.clearAllMocks)

    const getSamplePermission = ({
      licenceStartDate = LICENCE_START_DATE,
      licensee = getLicensee(67, licenceStartDate),
      concessions = []
    } = {}) => ({
      licenceStartDate,
      licensee,
      concessions: []
    })

    it('if the licensee is a minor, sets noLicenceRequired flag to true', () => {
      const licensee = getLicensee({ age: 8 })
      const permission = getSamplePermission({ licensee })
      ages.ageConcessionHelper(permission)
      expect(permission).toMatchSnapshot()
    })

    it('if the licensee is a junior, sets license data, junior concession and contact methods', () => {
      const licensee = getLicensee({ age: 14 })
      const permission = getSamplePermission({ licensee })
      ages.ageConcessionHelper(permission)
      expect(permission).toMatchSnapshot()
    })

    it('if the licensee is a senior, adds senior concession', () => {
      const licensee = getLicensee({ age: 67 })
      const permission = getSamplePermission({ licensee })
      ages.ageConcessionHelper(permission)
      expect(permission).toMatchSnapshot()
    })

    ;[junior, senior].forEach(concession => {
      it(`if is normal licence, removes ${concession.type} concession`, () => {
        const licensee = getLicensee({ age: 35 })
        const permission = getSamplePermission({ licensee, concessions: [concession] })
        ages.ageConcessionHelper(permission)
        expect(permission).toMatchSnapshot()
      })
    })
  })
})
