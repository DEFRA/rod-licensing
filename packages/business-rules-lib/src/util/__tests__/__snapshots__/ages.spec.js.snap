// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The concession helper ageConcessionHelper if is normal licence, removes Junior concession 1`] = `
Object {
  "concessions": Array [],
  "licenceStartDate": "2020-06-06",
  "licensee": Object {
    "birthDate": "1985-06-06",
    "firstName": "Johnny",
    "lastName": "Test",
    "preferredMethodOfConfirmation": "Letter",
  },
}
`;

exports[`The concession helper ageConcessionHelper if is normal licence, removes Senior concession 1`] = `
Object {
  "concessions": Array [],
  "licenceStartDate": "2020-06-06",
  "licensee": Object {
    "birthDate": "1985-06-06",
    "firstName": "Johnny",
    "lastName": "Test",
    "preferredMethodOfConfirmation": "Letter",
  },
}
`;

exports[`The concession helper ageConcessionHelper if the licensee is a junior, sets license data, junior concession and contact methods 1`] = `
Object {
  "concessions": Array [
    Object {
      "proof": Object {
        "type": "No Proof",
      },
      "type": "Junior",
    },
  ],
  "licenceLength": "12M",
  "licenceStartDate": "2020-06-06",
  "licenceStartTime": "0",
  "licensee": Object {
    "birthDate": "2006-06-06",
    "firstName": "Johnny",
    "lastName": "Test",
    "preferredMethodOfConfirmation": "Prefer not to be contacted",
    "preferredMethodOfReminder": "Prefer not to be contacted",
  },
}
`;

exports[`The concession helper ageConcessionHelper if the licensee is a minor, sets noLicenceRequired flag to true 1`] = `
Object {
  "licenceStartDate": "2020-06-06",
  "licensee": Object {
    "birthDate": "2012-06-06",
    "firstName": "Johnny",
    "lastName": "Test",
    "noLicenceRequired": true,
    "preferredMethodOfConfirmation": "Letter",
  },
}
`;

exports[`The concession helper ageConcessionHelper if the licensee is a senior, adds senior concession 1`] = `
Object {
  "concessions": Array [
    Object {
      "proof": Object {
        "type": "No Proof",
      },
      "type": "Senior",
    },
  ],
  "licenceStartDate": "2020-06-06",
  "licensee": Object {
    "birthDate": "1953-06-06",
    "firstName": "Johnny",
    "lastName": "Test",
    "preferredMethodOfConfirmation": "Letter",
  },
}
`;
